### PR TITLE
clean up BaseBlockTree::eraseBlock()

### PR DESCRIPTION
**Pull request by black0rwhite**

### CODE REVIEW

File: include/veriblock/pop/blockchain/base_block_tree.hpp [hunk 1]
```diff
@@ -519,6 +519,15 @@ struct BaseBlockTree {
     tips_.insert(index);
   }
 
+  //! dangerous! permanently drops the block from the tree
+  void eraseBlockUnchecked(index_t& block) {
```
Comment (Warchant) [inline include/veriblock/pop/blockchain/base_block_tree.hpp:523]: Maybe rename to deallocateBlockUnsafe?

Comment (blackyblack) [reply include/veriblock/pop/blockchain/base_block_tree.hpp:523]: Why make a separate function for erasing block? You can merge `eraseBlock` and `eraseBlockUnchecked` into a single function.

Comment (black0rwhite) [reply include/veriblock/pop/blockchain/base_block_tree.hpp:523]: neither of eraseBlock functions is particularly safe. erase sounds unambiguous enough to me. drop might be also work. deallocate is a bit too low-level/implementation detail.

File: include/veriblock/pop/blockchain/base_block_tree.hpp [hunk 2]
```diff
@@ -519,6 +519,15 @@ struct BaseBlockTree {
...
+    VBK_LOG_INFO("Deallocating block %s", block.toShortPrettyString());
+    auto count = blocks_.erase(makePrevHash(block.getHash()));
+    VBK_ASSERT_MSG(count == 1,
+                   "state corruption: block %s is not in the store",
+                   block.toPrettyString());
+  }
+
   //! dangerous! permanently drops the block from the tree
   void eraseBlock(index_t& block) {
```
Comment (Warchant) [inline include/veriblock/pop/blockchain/base_block_tree.hpp:532]: I am not convinced that Unsafe func is needed. It doesn't check that pnext is empty, which means that we can deallocate a block in a middle of a chain. 
I'd remove Unsafe version.

**Review by Warchant — CHANGES_REQUESTED** [review]

File: include/veriblock/pop/blockchain/base_block_tree.hpp (hunk 2 above)
Comment (black0rwhite) [reply include/veriblock/pop/blockchain/base_block_tree.hpp:532]: This is a refactoring commit. so, all it is intended to do is deduplicate code and regularize logging. 
Whether this function has a future depends on what happens to eraseBlock. We either want eraseLeaf+eraseRoot combo or a generalized eraseBlock that ensures we can only delete a root or a leaf. eraseLeaf/Root will have a use for eraseBlockUnchecked and I personally lean towards this approach as it is more explicit

File: include/veriblock/pop/blockchain/base_block_tree.hpp (hunk 1 above)
Comment (black0rwhite) [reply include/veriblock/pop/blockchain/base_block_tree.hpp:523]: one more point regarding "deallocate": what would you call this function in java?

Comment (black0rwhite) [reply include/veriblock/pop/blockchain/base_block_tree.hpp:523]: > Why make a separate function for erasing block? You can merge eraseBlock and eraseBlockUnchecked into a single function.
The point of multiple erase functions is contracts/invariant enforcement via asserts.

Comment (blackyblack) [reply include/veriblock/pop/blockchain/base_block_tree.hpp:523]: > 

Well, eraseBlock should be eraseLeafBlock then. And eraseBlockUnchecked should be eraseAnyBlock.

Comment (black0rwhite) [reply include/veriblock/pop/blockchain/base_block_tree.hpp:523]: eraseBlock can erase the root block if the tree contains a single block, it's a bit of a mess atm. eraseBlockUnchecked was intended to sound unsafe as compared to eraseBlock.

Comment (blackyblack) [reply include/veriblock/pop/blockchain/base_block_tree.hpp:523]: If the tree contains a single block then root block is also a leaf block.

Comment (black0rwhite) [reply include/veriblock/pop/blockchain/base_block_tree.hpp:523]: hmm, although we don't explicitly check that the block is not the root, we assert that block.isDeleted(), so in a tree that isn't malformed, we can't use the current implementation of eraseBlock() to erase the root block.

Comment (black0rwhite) [reply include/veriblock/pop/blockchain/base_block_tree.hpp:523]: anyway, this commit wasn't intended to make erase* stuff perfect: https://github.com/VeriBlock/alt-integration-cpp/pull/898#discussion_r759164702

Comment (black0rwhite) [reply include/veriblock/pop/blockchain/base_block_tree.hpp:523]: As to whether the root block is a leaf block, yes, formally it is. However, deleting the root block is not something the developer expects when manipulating leaf blocks. eraseRoot/eraseLeaf split is formally eraseRoot/eraseLeafThatIsNotRoot, but that's a bit too noisy.

**Comment by Warchant** [comment]
I am not convinced that this change is necessary. I am open for discussion if you think otherwise.

**Pull request closed without merging**